### PR TITLE
Fix threadsafety of Dynamoid::Adapter (issue #10)

### PR DESCRIPTION
**Pull request by philipmw**
I changed Dynamoid::Adapter from being a module having global state to a class
that's instantiated by Dynamoid module.  This makes Adapter more testable and
allows me to write specs verifying its lazy-connection behavior.

I extend the lazy behavior to caching tables -- instead of caching tables when
establishing a connection, we cache on first demand.

Turning Dynamoid::Adapter into a class required me to rename the `Adapter`
namespace (module) of AwsSdkV2 to `AdapterPlugin`.

### CODE REVIEW

**Comment by philipmw** [comment]
@loganb Can I get your eyes on it?

File: lib/dynamoid/adapter.rb [hunk 1]
```diff
@@ -5,27 +5,48 @@ module Dynamoid
...
+      @tables_mutex = Mutex.new
     end
 
-    # The actual adapter currently in use: presently AwsSdk.
-    #
-    # @since 0.2.0
-    def self.adapter
-      reconnect! unless @adapter
-      @adapter
+    def tables
```
Comment (loganb) [inline lib/dynamoid/adapter.rb]: So the double-check locking idiom is (mostly) not safe. In very rare circumstances, I think this code can fail. See: http://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html

File: lib/dynamoid/adapter_plugin/aws_sdk_v2.rb [hunk 2]
```diff
@@ -1,5 +1,5 @@
 module Dynamoid
-  module Adapter
+  module AdapterPlugin
```
Comment (loganb) [inline lib/dynamoid/adapter_plugin/aws_sdk_v2.rb]: So how does this work given that the class name doesn't match the file name?

**Comment by loganb** [comment]
I like the move away from global state and towards instantiated classes. Do you think you can avoid locking entirely by eagerly instantiating the client and assigning it to a variable? If not, you may want to consider using an atomic variable rather than a mutex, or just always lock when retrieving the adapter.

File: lib/dynamoid/adapter.rb (hunk 1 above)
Comment (philipmw) [inline lib/dynamoid/adapter.rb]: Ok, I will switch to something else.

File: lib/dynamoid/adapter_plugin/aws_sdk_v2.rb (hunk 2 above)
Comment (philipmw) [inline lib/dynamoid/adapter_plugin/aws_sdk_v2.rb]: They do match... the module name is `AdapterPlugin` while its directory is `adapter_plugin`, and the class name is unchanged.

**Comment by philipmw** [comment]
Ok, I'll see what I can do. Thanks for the feedback.

File: lib/dynamoid/adapter_plugin/aws_sdk_v2.rb (hunk 2 above)
Comment (loganb) [inline lib/dynamoid/adapter_plugin/aws_sdk_v2.rb]: Whoops! My bad. :)

**Comment by philipmw** [comment]
This is ready for another review from someone.

**Comment by philipmw** [comment]
This got auto-closed because I merged (and subsequently deleted) a branch atop which this PR is based.

**Pull request closed without merging**